sympy/sets/tests/test_sets.py

```diff
@@ -1597,6 +1600,42 @@ def test_DisjointUnion_len():
     assert len(DisjointUnion(S.EmptySet, S.EmptySet, FiniteSet(x, y, z), S.EmptySet)) == 3
     raises(ValueError, lambda: len(DisjointUnion(Interval(0, 1), S.EmptySet)))
 
+def test_SetKind_ProductSet():
+    p = ProductSet(FiniteSet(Matrix([1, 2])), FiniteSet(Matrix([1, 2])))
+    mk = MatrixKind(NumberKind)
+    k = SetKind(TupleKind(mk, mk))
+    assert p.kind is k
+    assert ProductSet(Interval(1, 2), FiniteSet(Matrix([1, 2]))).kind is SetKind(TupleKind(NumberKind, mk))
+
+def test_SetKind_Interval():
+    assert Interval(1, 2).kind is SetKind(NumberKind)
+
+def test_SetKind_EmptySet_UniversalSet():
+    assert S.UniversalSet.kind is SetKind(UndefinedKind)
+    assert EmptySet.kind is SetKind()
+
+def test_SetKind_FiniteSet():
+    assert FiniteSet(1, Matrix([1, 2])).kind is SetKind(UndefinedKind)
+    assert FiniteSet(1, 2).kind is SetKind(NumberKind)
+
+def test_SetKind_Unions():
+    assert Union(FiniteSet(Matrix([1, 2])), Interval(1, 2)).kind is SetKind(UndefinedKind)
+    assert Union(Interval(1, 2), Interval(1, 7)).kind is SetKind(NumberKind)
+
+def test_SetKind_evaluate_False():
+    one = Union({1}, EmptySet, evaluate=False).kind
+    assert one is SetKind(NumberKind)
+    two = Union(Interval(1, 2), EmptySet, evaluate=False).kind
+    assert two is SetKind(NumberKind)
+    three = Intersection({1}, S.UniversalSet,evaluate=False).kind
+    assert three is SetKind(NumberKind)
+    four = Union({1}, S.UniversalSet, evaluate=False).kind
+    assert four is SetKind(UndefinedKind)
+    five = Union(Interval(1, 2), Interval(4, 5), FiniteSet(1), evaluate=False).kind
+    assert five is SetKind(NumberKind)
+    six = Intersection({1}, EmptySet, evaluate=False).kind
+    assert six is SetKind()
```

Comment:
The idea should be to make it as easy as possible to read the different examples. How about:
```python
U = lambda *args: Union(*args, evaluate=False)
assert U({1}, EmptySet).kind = SetKind(NumberKind)
assert U(Interval(1, 2), EmptySet).kind is SetKind(NumerKind)
...
```